servlet-core/src/main/java/io/micronaut/servlet/http/DefaultBodyBuilder.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2017-2023 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.servlet.http;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.AnnotationMetadata;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.io.buffer.ByteBuffer;
+import io.micronaut.core.type.Argument;
+import io.micronaut.core.util.CollectionUtils;
+import io.micronaut.http.HttpAttributes;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.http.MediaType;
+import io.micronaut.http.annotation.Body;
+import io.micronaut.http.codec.CodecException;
+import io.micronaut.http.codec.MediaTypeCodec;
+import io.micronaut.http.codec.MediaTypeCodecRegistry;
+import io.micronaut.inject.ExecutionHandle;
+import io.micronaut.json.codec.MapperMediaTypeCodec;
+import io.micronaut.json.tree.JsonNode;
+import io.micronaut.web.router.RouteMatch;
+import jakarta.inject.Singleton;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+/**
+ * Default implementation of {@link BodyBuilder}.
+ * @author Sergio del Amo
+ * @since 4.0.0
+ */
+@Requires(classes = {JsonNode.class, MapperMediaTypeCodec.class})
+@Internal
+@Singleton
+public class DefaultBodyBuilder implements BodyBuilder {
+    private static final Set<Class<?>> RAW_BODY_TYPES = CollectionUtils.setOf(String.class, byte[].class, ByteBuffer.class, InputStream.class);
+
+    private final MediaTypeCodecRegistry codecRegistry;
+
+    public DefaultBodyBuilder(MediaTypeCodecRegistry codecRegistry) {
+        this.codecRegistry = codecRegistry;
+    }
+
+    @Override
+    @Nullable
+    public Object buildBody(@NonNull Callable<InputStream> bodySupplier,
+                             @NonNull HttpRequest<?> request) {
+        final MediaType contentType = request.getContentType().orElse(MediaType.APPLICATION_JSON_TYPE);
```

Comment:
This should likely use the new `MessageBodyReader` API not not be coupled to JSON

//cc @yawkat

Reply:
agreed

Reply:
@yawkat is this change to use the `MessageBodyReader`API  something you can do?

Reply:
As discussed I will create an issue to do this afterwards.